File: reference/moneymapper_app/lib/services/auth_service.dart
import 'package:shared_preferences/shared_preferences.dart';
import 'package:supabase_flutter/supabase_flutter.dart';
import 'resilience_utils.dart';

class AuthService {
  static const _tokenKey = 'jwt_token';
  static const _nameKey = 'user_name';
  static const _emailKey = 'user_email';
  static const _planKey = 'user_plan';
  static const _profileCacheKey = 'master_profile_data';

  final SupabaseClient _supabase = Supabase.instance.client;

  Future<void> saveToken(String token) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_tokenKey, token);
  }

  Future<String?> getToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_tokenKey);
  }

  Future<void> saveUser(Map<String, dynamic> user) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_nameKey, user['name']?.toString() ?? '');
    await prefs.setString(_emailKey, user['email']?.toString() ?? '');
    final currentPlan = prefs.getString(_planKey);
    if (currentPlan == null || currentPlan.isEmpty) {
      await prefs.setString(_planKey, user['plan']?.toString() ?? 'b2c');
    }
  }

  Future<void> savePlan(String plan) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_planKey, plan);
  }

  Future<String?> getUserName() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_nameKey);
  }

  Future<String?> getUserEmail() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_emailKey);
  }

  Future<String?> getUserPlan() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_planKey);
  }

  Future<void> logout() async {
    try {
      await _supabase.auth.signOut();
    } catch (e) {
      // Local cleanup even if network sign-out fails
    }
    final prefs = await SharedPreferences.getInstance();
    await prefs.clear();
  }

  Future<bool> isLoggedIn() async {
    return _supabase.auth.currentSession != null;
  }

  // --- Profile Caching Methods ---

  Future<void> saveMasterProfileLocally(String jsonStr) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_profileCacheKey, jsonStr);
  }

  Future<String?> getMasterProfileLocally() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_profileCacheKey);
  }

  // --- Resilient Auth Methods ---

  Future<AuthResponse> signUp({
    required String email,
    required String password,
    required Map<String, dynamic> metadata,
  }) async {
    try {
      return await ResilienceUtils.retry(
        context: "Auth_SignUp",
        task: () => _supabase.auth.signUp(email: email, password: password, data: metadata),
      );
    } on AuthException catch (e) {
      if (e.message.contains('already registered')) throw 'An account already exists with this email.';
      throw ResilienceUtils.sanitizeErrorMessage(e);
    }
  }

  Future<AuthResponse> signIn({
    required String email,
    required String password,
  }) async {
    try {
      final response = await ResilienceUtils.retry(
        context: "Auth_SignIn",
        task: () => _supabase.auth.signInWithPassword(email: email, password: password),
      );

      if (response.user != null) {
        final metadata = response.user!.userMetadata ?? {};
        await saveUser({
          'name': metadata['fullName'] ?? '',
          'email': response.user!.email ?? '',
          'plan': metadata['plan'] ?? 'b2c',
        });
        if (response.session != null) {
          await saveToken(response.session!.accessToken);
        }
      }
      return response;
    } on AuthException catch (e) {
      if (e.message.contains('Invalid login credentials')) {
        throw 'Incorrect email or password. Please try again.';
      }
      throw ResilienceUtils.sanitizeErrorMessage(e);
    }
  }

  Future<void> resetPassword(String email) async {
    try {
      await _supabase.auth.resetPasswordForEmail(email, redirectTo: 'moneymapper://reset-callback');
    } catch (e) {
      throw ResilienceUtils.sanitizeErrorMessage(e);
    }
  }

  Future<void> updatePassword(String newPassword) async {
    try {
      await _supabase.auth.updateUser(UserAttributes(password: newPassword));
    } catch (e) {
      throw ResilienceUtils.sanitizeErrorMessage(e);
    }
  }

  String? get currentUserId => _supabase.auth.currentUser?.id;

  Future<void> syncMetadata() async {
    final user = _supabase.auth.currentUser;
    if (user != null) {
      final prefs = await SharedPreferences.getInstance();
      final existingPlan = prefs.getString(_planKey);

      final metadata = user.userMetadata ?? {};
      await saveUser({
        'name': metadata['fullName'] ?? '',
        'email': user.email ?? '',
        'plan': existingPlan ?? metadata['plan'] ?? 'b2c',
      });
    }
  }
}

File: src/services/authService.ts
import { supabase } from '../lib/supabase';
import type { User, AuthResponse } from '@supabase/supabase-js';

export interface SignUpMetadata {
  fullName: string;
  mobile?: string;
  plan?: string;
  [key: string]: any;
}

export interface StoredUser {
  name?: string;
  email?: string;
  plan?: string;
}

export const AUTH_STORAGE_KEYS = {
  NAME: 'user_name',
  EMAIL: 'user_email',
  PLAN: 'user_plan',
  PROFILE_CACHE: 'master_profile_data',
} as const;

/**
 * Sanitizes technical error messages for end-users, mirroring Flutter ResilienceUtils.
 */
export function sanitizeErrorMessage(error: unknown): string {
  let rawMsg = '';
  if (error instanceof Error) {
    rawMsg = error.message;
  } else if (typeof error === 'object' && error !== null && 'message' in error) {
    rawMsg = String((error as any).message);
  } else {
    rawMsg = String(error);
  }
  const msg = rawMsg.toLowerCase();

  if (
    msg.includes('network') ||
    msg.includes('socket') ||
    msg.includes('failed host') ||
    msg.includes('fetch')
  ) {
    return 'Connection issue detected. Please check your internet and try again.';
  }
  if (msg.includes('429') || msg.includes('rate limit')) {
    return "Servers are busy right now. We'll refresh your data in a moment.";
  }
  if (msg.includes('timeout')) {
    return 'The request took too long. Please try refreshing again.';
  }
  if (msg.includes('auth') || msg.includes('401') || msg.includes('403')) {
    return 'Session expired or unauthorized. Please log in again.';
  }

  return "Something went wrong on our end. We've logged this and are looking into it.";
}

export class AuthService {
  private _currentUser: User | null = null;

  setCurrentUser(user: User | null): void {
    this._currentUser = user;
  }

  get currentUserId(): string | null {
    if (this._currentUser?.id) {
      return this._currentUser.id;
    }
    // Synchronous fallback by reading Supabase persisted session from localStorage
    if (typeof window !== 'undefined' && window.localStorage) {
      try {
        for (let i = 0; i < localStorage.length; i++) {
          const key = localStorage.key(i);
          if (key && key.startsWith('sb-') && key.endsWith('-auth-token')) {
            const raw = localStorage.getItem(key);
            if (raw) {
              const parsed = JSON.parse(raw);
              const id = parsed?.user?.id;
              if (id) return id;
            }
          }
        }
      } catch {
        // ignore parse error
      }
    }
    return null;
  }

  async saveUser(user: StoredUser): Promise<void> {
    if (typeof window === 'undefined' || !window.localStorage) return;
    localStorage.setItem(AUTH_STORAGE_KEYS.NAME, user.name ?? '');
    localStorage.setItem(AUTH_STORAGE_KEYS.EMAIL, user.email ?? '');
    const currentPlan = localStorage.getItem(AUTH_STORAGE_KEYS.PLAN);
    if (!currentPlan) {
      localStorage.setItem(AUTH_STORAGE_KEYS.PLAN, user.plan ?? 'b2c');
    }
  }

  async savePlan(plan: string): Promise<void> {
    if (typeof window === 'undefined' || !window.localStorage) return;
    localStorage.setItem(AUTH_STORAGE_KEYS.PLAN, plan);
  }

  async getUserName(): Promise<string | null> {
    if (typeof window === 'undefined' || !window.localStorage) return null;
    return localStorage.getItem(AUTH_STORAGE_KEYS.NAME);
  }

  async getUserEmail(): Promise<string | null> {
    if (typeof window === 'undefined' || !window.localStorage) return null;
    return localStorage.getItem(AUTH_STORAGE_KEYS.EMAIL);
  }

  async getUserPlan(): Promise<string | null> {
    if (typeof window === 'undefined' || !window.localStorage) return null;
    return localStorage.getItem(AUTH_STORAGE_KEYS.PLAN);
  }

  async saveMasterProfileLocally(jsonStr: string): Promise<void> {
    if (typeof window === 'undefined' || !window.localStorage) return;
    localStorage.setItem(AUTH_STORAGE_KEYS.PROFILE_CACHE, jsonStr);
  }

  async getMasterProfileLocally(): Promise<string | null> {
    if (typeof window === 'undefined' || !window.localStorage) return null;
    return localStorage.getItem(AUTH_STORAGE_KEYS.PROFILE_CACHE);
  }

  /**
   * Clears app-owned keys and cached items from localStorage.
   * Explicitly preserves 'theme_mode' and other unrelated items.
   */
  clearLocalData(): void {
    if (typeof window === 'undefined' || !window.localStorage) return;

    // Remove static app keys
    localStorage.removeItem(AUTH_STORAGE_KEYS.NAME);
    localStorage.removeItem(AUTH_STORAGE_KEYS.EMAIL);
    localStorage.removeItem(AUTH_STORAGE_KEYS.PLAN);
    localStorage.removeItem(AUTH_STORAGE_KEYS.PROFILE_CACHE);

    // Remove dynamic cache keys and sb-*-auth-token entries
    const keysToRemove: string[] = [];
    for (let i = 0; i < localStorage.length; i++) {
      const key = localStorage.key(i);
      if (key) {
        if (
          key.startsWith('dashboard_') ||
          key.startsWith('profile_') ||
          key.startsWith('weekly_') ||
          key.startsWith('corp_stats_') ||
          key.startsWith('mm_cache_') ||
          key.startsWith('mm_cache_ts_') ||
          (key.startsWith('sb-') && key.endsWith('-auth-token'))
        ) {
          keysToRemove.push(key);
        }
      }
    }
    keysToRemove.forEach((key) => localStorage.removeItem(key));
  }

  /**
   * Logs out the user.
   * Calls signOut, falls back to local signOut on failure, and ensures all local
   * session and app cache data is cleared even if network fails.
   */
  async logout(clientOverride?: any): Promise<void> {
    const client = clientOverride || supabase;
    try {
      const res = await client.auth.signOut();
      if (res?.error) {
        await client.auth.signOut({ scope: 'local' }).catch(() => {});
      }
    } catch {
      try {
        await client.auth.signOut({ scope: 'local' });
      } catch {
        // ignore
      }
    } finally {
      this._currentUser = null;
      this.clearLocalData();
    }
  }

  async isLoggedIn(): Promise<boolean> {
    try {
      const { data } = await supabase.auth.getSession();
      return !!data?.session;
    } catch {
      return false;
    }
  }

  async signUp(email: string, password: string, metadata: SignUpMetadata): Promise<AuthResponse['data']> {
    let res: AuthResponse;
    try {
      res = await supabase.auth.signUp({
        email,
        password,
        options: {
          data: {
            plan: 'b2c',
            ...metadata,
          },
        },
      });
    } catch (err: any) {
      const msg = String(err?.message || err);
      if (msg.toLowerCase().includes('already registered')) {
        throw new Error('An account already exists with this email.');
      }
      throw new Error(sanitizeErrorMessage(err));
    }

    const { data, error } = res;
    if (error) {
      const errorMsg = error.message || '';
      if (errorMsg.toLowerCase().includes('already registered')) {
        throw new Error('An account already exists with this email.');
      }
      throw new Error(sanitizeErrorMessage(error));
    }

    if (data.user) {
      this._currentUser = data.user;
    }

    return data;
  }

  async signIn(email: string, password: string): Promise<AuthResponse['data']> {
    let res: AuthResponse;
    try {
      res = await supabase.auth.signInWithPassword({
        email,
        password,
      });
    } catch (err: any) {
      const msg = String(err?.message || err);
      if (msg.toLowerCase().includes('invalid login credentials')) {
        throw new Error('Incorrect email or password. Please try again.');
      }
      throw new Error(sanitizeErrorMessage(err));
    }

    const { data, error } = res;
    if (error) {
      const errorMsg = error.message || '';
      if (errorMsg.toLowerCase().includes('invalid login credentials')) {
        throw new Error('Incorrect email or password. Please try again.');
      }
      throw new Error(sanitizeErrorMessage(error));
    }

    if (data.user) {
      this._currentUser = data.user;
      const metadata = data.user.user_metadata || {};
      await this.saveUser({
        name: metadata.fullName ?? '',
        email: data.user.email ?? '',
        plan: metadata.plan ?? 'b2c',
      });
    }

    return data;
  }

  async resetPassword(email: string): Promise<{}> {
    const origin = typeof window !== 'undefined' ? window.location.origin : '';
    const redirectTo = `${origin}/reset-password`;
    let res: any;

    try {
      res = await supabase.auth.resetPasswordForEmail(email, {
        redirectTo,
      });
    } catch (err: any) {
      throw new Error(sanitizeErrorMessage(err));
    }

    const { data, error } = res;
    if (error) {
      throw new Error(sanitizeErrorMessage(error));
    }

    return data;
  }

  async updatePassword(newPassword: string): Promise<{ user: User | null }> {
    let res: any;
    try {
      res = await supabase.auth.updateUser({
        password: newPassword,
      });
    } catch (err: any) {
      throw new Error(sanitizeErrorMessage(err));
    }

    const { data, error } = res;
    if (error) {
      throw new Error(sanitizeErrorMessage(error));
    }

    if (data.user) {
      this._currentUser = data.user;
    }

    return data;
  }

  async syncMetadata(): Promise<void> {
    try {
      const { data } = await supabase.auth.getUser();
      const user = data?.user ?? this._currentUser;
      if (user) {
        const existingPlan = typeof window !== 'undefined' ? localStorage.getItem(AUTH_STORAGE_KEYS.PLAN) : null;
        const metadata = user.user_metadata || {};
        await this.saveUser({
          name: metadata.fullName ?? '',
          email: user.email ?? '',
          plan: existingPlan || metadata.plan || 'b2c',
        });
      }
    } catch {
      // ignore sync failure
    }
  }
}

export const authService = new AuthService();
